feat: add latest partition support for BigQuery

mistercrunch opened the pull request:
Adding db_engine_spec-related features that enables SQL Lab to show the latest partition when using time partitioning in BigQuery as well as applying a WHERE clause on the latest partition by default when fetching the sample dataset. Turns out that `SELECT * FROM {{...}} LIMIT {n}` can be costly against large tables in BigQuery as it results in a full table scan.

closes https://github.com/apache/superset/issues/17299

mistercrunch committed:
feat(sqllab): add latest partition support for BigQuery

Adding db_engine_spec-related features that enables SQL Lab to show the latest partition when using time partitioning in BigQuery as well as applying a WHERE clause on the latest partition by default when fetching the sample dataset. Turns out that `SELECT * FROM {{...}} LIMIT {n}` can be costly against large tables in BigQuery as it results in a full table scan.

mistercrunch committed:
formating

mistercrunch committed:
fix tests

mistercrunch commented on tests/integration_tests/db_engine_specs/bigquery_tests.py:
```diff
@@ -111,108 +121,45 @@ def values(self):
-        )
-        assert result == expected_result
-
-    def test_get_indexes(self):
```
turns out BigQuery's implementation of get_indexes is a no-op -> https://github.com/googleapis/python-bigquery-sqlalchemy/blob/150143ce27c02f1cd711b8df4ffe587a98ada288/sqlalchemy_bigquery/base.py#L1266-L1268

betodealmeida replied:
#TIL

mistercrunch committed:
minor fix

rusackas commented on docker/pythonpath_dev/superset_config.py:
```diff
@@ -40,6 +40,7 @@
 EXAMPLES_HOST = os.getenv("EXAMPLES_HOST")
 EXAMPLES_PORT = os.getenv("EXAMPLES_PORT")
 EXAMPLES_DB = os.getenv("EXAMPLES_DB")
+SHOW_STACKTRACE = True
```
Was this intended to be part of the PR, or just for local/temporary dev purposes?

mistercrunch replied:
Let me remove that line, now thinking it may make sense to deprecate this flag as it's not evenly implemented.

mistercrunch committed:
rm SHOW_STACKTRACES

mistercrunch committed:
keep normalize_indexes in base

betodealmeida commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -284,66 +290,90 @@ def _truncate_label(cls, label: str) -> str:
+            FROM `{table.schema}.INFORMATION_SCHEMA.PARTITIONS`
+            WHERE table_name = '{table.table}'
+        """)
+        df = database.get_df(sql)
```
```suggestion
        df = database.get_df(sql, catalog=table.catalog, schema=table.schema)
```

betodealmeida replied:
I think we need a standard method in the DB engine spec to run SQL. Looks like some engine specs use `engine.execute`, some use `cursor.execute`, some `inspector.bind.execute`, and others `database.get_df`. Something where `catalog` and `schema` are not optional (but can be null) to prevent people from forgetting to pass them:

```
class BaseEngineSpec:
    def execute_sql(self, sql: str, catalog: str | None, schema: str | None):
        ...
```

betodealmeida commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -284,66 +290,90 @@ def _truncate_label(cls, label: str) -> str:
+            database, catalog=table.catalog, schema=table.schema
+        ) as engine:
+            client = cls._get_client(engine, database)
+            bq_table = client.get_table(f"{table.schema}.{table.table}")
```
Here we don't need catalog because `engine` is already bound to it.

betodealmeida reviewed (commented):
@mistercrunch can you test with "allow changing catalogs enabled"? I think we need to pass the catalog in a few places.

mistercrunch committed:
fully qualified table name

mistercrunch commented:
> can you test with "allow changing catalogs enabled"?

Hit an unrelated UI bug trying to edit the json object, reported on internal slack

viyh commented:
@mistercrunch, any chance of getting this PR fixed/merged soon? This is a big deal for my team as we make heavy use of BigQuery and LIMIT is just not safe. Appreciate it, thanks!

mistercrunch committed:
avoid sql injection

mistercrunch commented:
@viyh I just updated the PR:
- rebased on latest master
- fixed the potential SQL injection issue using the sqla sql toolkit
- @betodealmeida I tested with and without the "allow catalog" checkbox

Should be go for another round of review and hopefully merging.

One thing I noticed - for another PR, is that we don't seem to get the indexes here anymore. The driver has changed and they made get_indexes a no-op now. Something to look at. This current PR is much more critical given the current state of doing a "full scan" on preview

mistercrunch commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -289,66 +295,115 @@ def _truncate_label(cls, label: str) -> str:
+        ).where(partitions_table.c.table_name == table.table)
+
+        # Compile to BigQuery SQL
+        compiled_query = query.compile(
```
there may be a better way to do this. Felt like `dialect` should be in-context here

viyh commented:
@mistercrunch 🙏 Thank you! Although looks like it failed some different checks now, naturally. :)

mistercrunch committed:
late import in case it doesn't exist

mistercrunch committed:
use Database.get_dialect()

mistercrunch commented:
@viyh  I think I just fixed it, found a better way to retrieve the dialect with `Database.get_dialect()`

betodealmeida commented on superset/db_engine_specs/bigquery.py:
```diff
@@ -289,66 +294,115 @@ def _truncate_label(cls, label: str) -> str:
+        if df.empty or df.iat[0, 0] is None:
+            return None
+
+        return df.iat[0, 0]
```
Compiling and loading into a dataframe seems a bit overkill? Why not something like:

```python
          with database.get_raw_connection(
              catalog=table.catalog,
              schema=table.schema,
          ) as conn:
              cursor = conn.cursor()
              cursor.execute("""
                  SELECT MAX(partition_id)
                  FROM INFORMATION_SCHEMA.PARTITIONS
                  WHERE table_name=%s
              """, (table.table,))
              if row := cursor.fetchone():
                  return row[0]

              return None
```

We should standardize how DB engine specs run queries, BTW! 😄

mistercrunch replied:
> We should standardize how DB engine specs run queries, BTW! 

Totally, I just used `get_df` cause it's an easy way to run a query...

betodealmeida reviewed (approved):
Looks good, just left a suggestion about using a cursor instead of `get_df(sql)`.

mistercrunch committed:
addressing comments
